### ISSUE TITLE
[BUG] Slow performance of multiple item selection

### ISSUE THREAD

**Issue by mneuwert** [owncloud/ios-app#735]
## If you found a bug:
In the file list containing big number of items, the more items are selected, the slower selection / deselection of new items becomes.

### Steps to reproduce
1. Make sure you have a folder with say 200-300 items
2. Select All
3. Try to deselect / select again some items

### Expected behavior
UI should not freeze and items should be selected / deselected without noticeable lag

### Actual behavior
It takes time and UI freezes, so you can't scroll the list for a second or two depending on how capable the iOS device is.

### Client
**iOS version: 13.5.1**

**ownCloud app version: 11.4**

**Device model: iPhone 11 Pro**

### Background
The performance bottle-neck is caused by implementation of multiple-selection in `ClientQueryViewController` class:
- There is a loop running on main thread running over all selected items and then matching them against actions available for the selected collection of items. This is done to make sure that only actions which are allowed to be performed are enabled (e.g. if at least one item in the selected list doesn't have 'delete' permission or is root of the share, then delete action is disabled)
- Action matching mechanism (see `Action` / `ActionContext`) implementation is actually requiring the array of items to check if particular action is available for those items (e.g. delete, duplicate etc). If we change this, we can improve performance dramatically.
- So instead of matching actions against whole list of items on each selection / deselection, we could just keep track of reference counted shared set of few properties for the list of selected items including e.g. item permissions, information about shared items.

**Example:**
Instead of iterating the list and trying to find a single item which doesn't allow certain action, you could have a counted set which keep tracks of how many items are writeable, deletable etc. etc.

On selection, you would increment counters of permissions according to what permissions are set in this item.

On deselection, you would decrement the counters.

**Comment by felix-schwarz** [comment]
Thanks for highlighting a performance bottleneck. Did you benchmark which actions takes the longest to match?

F.ex. some actions are only available for a single item and will only check on the number of items and return immediately if they're above 1.

What I'd therefore suggest is not a change of the whole mechanism, but an extension to it: if we add a dictionary to the `ActionContext` with the summary/counter information, actions that only need that information can return a matching result quickly, based on that information. But if we have an action that needs to really check each item individually, that possibility is also still there and available.

**Comment by mneuwert** [comment]
@felix-schwarz My idea was also to extend `ActionContext`... I was sort of thinking about it as counted set (e.g. [ writeable(5), readable(5), deletable(4), shared(1) ... ]) instead of dictionary but it could be also a dictionary which is update every time new `OCItem` gets selected / deselected.

I did some quick profiling and found e.g. that for instance following methods are eating up significant chunk of CPU time:

Almost 83% percent of burned CPU cycles account to`Action.sortedApplicableActions(for context:)` and then there are few hot spots like:
- `DeleteAction.applicablePosition(forContext:)`
- `MoveAction.applicablePosition(forContext:)`
- `OpenInAction.applicablePosition(forContext:)`

This boils down to `filter()`, `contains()` operations performed on a huge array of `OCItem` instances.

**Other ideas:**
- Consider replacing toolbar icons which are anyway sometimes hard to hit with a single more icon bringing up familiar card interface which is also not limited to few toolbar actions. This will not yet solve slow matching of actions for big item count though. However matching would be performed only once and not on every item selection / deselection. And you user is not limited in what he can do by the toolbar size. So right now e.g. you can't favourite multiple selected files.

- Relax matching rules for critical actions to make matching process quicker and post-pone time-critical checks until the action actually runs: e.g. if you select few thousands of files in macOS Finder UI and hit delete, it won't do sort of preflight operation and you'll just get errors along the way when delete operation is being executed.

- Match actions lazily: update a data-structure (dictionary, counted set) describing current working set and re-match actions only if it changes in a way which would affect matching results.
